File: collector/redactor.py
"""Secret redaction module, shared logic with the C# server-side redactor
described in 03-sicurezza-threat-model.md (section 6).

Pure stdlib, no I/O, safe to import at any time. The single public entry
point is redact(text) -> str: it replaces every recognized secret-shaped
substring with a fixed placeholder "[REDACTED:<category>]" and returns the
result. It never raises on malformed input (empty string, None-like input
is not expected but falsy input is handled defensively).

Categories (see 03 section 6):
  - token        : known provider token prefixes + generic key=value/bearer patterns
  - private_key  : PEM private key blocks
  - conn_string  : URI connection strings with embedded credentials
  - healthcheck  : healthcheck ping URLs (e.g. hc-ping.com/<uuid>-style paths)
  - internal_host: private IPv4 addresses and internal-looking hostnames
  - entropy      : fallback heuristic for high-entropy strings that did not
                    match any known pattern (fail-safe: prefer a false
                    positive redaction over a false negative exposure)
"""

from __future__ import annotations

import math
import re

PLACEHOLDER = "[REDACTED:{category}]"

# --- Known / generic tokens ---------------------------------------------

# Recognizable provider token prefixes (GitHub PAT family, fine-grained PAT).
_KNOWN_TOKEN_PREFIXES = re.compile(
    r"\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9]{20,255}\b"
    r"|\bgithub_pat_[A-Za-z0-9_]{20,255}\b"
)

# Generic "token=", "key=", "authorization:", "bearer " style assignments.
# Captures the key/prefix separately so it is preserved and only the value
# is redacted.
_GENERIC_TOKEN_ASSIGNMENT = re.compile(
    r"(?i)\b(token|api[_-]?key|secret|password|passwd|pwd|authorization)\s*[:=]\s*"
    r"['\"]?([A-Za-z0-9\-_./+=]{8,})['\"]?"
)

_BEARER_HEADER = re.compile(r"(?i)\bbearer\s+([A-Za-z0-9\-_.+/=]{8,})")

# --- PEM private keys -----------------------------------------------------

_PEM_PRIVATE_KEY = re.compile(
    r"-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----.*?-----END [A-Z0-9 ]*PRIVATE KEY-----",
    re.DOTALL,
)

# --- Connection strings ----------------------------------------------------

# scheme://user:password@host[:port]/... — any scheme (postgres, redis, etc).
_CONN_STRING = re.compile(
    r"\b[a-zA-Z][a-zA-Z0-9+.\-]*://[^\s:/@]+:[^\s@/]+@[^\s/'\"]+"
)

# --- Healthcheck ping URLs --------------------------------------------------

# healthchecks.io-style ping URLs: https://hc-ping.com/<uuid> or a self-hosted
# equivalent path ending in a UUID-shaped identifier used as a bearer token.
_HEALTHCHECK_URL = re.compile(
    r"(?i)\bhttps?://[^\s'\"]*\bhc-ping\b[^\s'\"]*"
    r"|\bhttps?://[^\s'\"]+/ping/[0-9a-f-]{16,}[^\s'\"]*"
)

# --- Private IPs / internal hostnames --------------------------------------

_PRIVATE_IPV4 = re.compile(
    r"\b(?:"
    r"10(?:\.\d{1,3}){3}"
    r"|172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2}"
    r"|192\.168(?:\.\d{1,3}){2}"
    # CGNAT 100.64.0.0/10 — used by mesh VPNs such as Tailscale.
    r"|100\.(?:6[4-9]|[7-9]\d|1[01]\d|12[0-7])(?:\.\d{1,3}){2}"
    r"|127(?:\.\d{1,3}){3}"
    r")\b"
)

# Internal-looking hostnames: single-label or *.local / *.internal / *.lan
# suffixes. Conservative on purpose to avoid redacting public domains.
_INTERNAL_HOSTNAME = re.compile(
    r"\b[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.(?:local|internal|lan|home)\b"
)

# --- Entropy heuristic -------------------------------------------------------

_CANDIDATE_TOKEN = re.compile(r"\b[A-Za-z0-9+/_=-]{20,}\b")
_ENTROPY_THRESHOLD_BITS_PER_CHAR = 3.5


def _shannon_entropy(s: str) -> float:
    if not s:
        return 0.0
    counts: dict[str, int] = {}
    for ch in s:
        counts[ch] = counts.get(ch, 0) + 1
    length = len(s)
    entropy = 0.0
    for count in counts.values():
        p = count / length
        entropy -= p * math.log2(p)
    return entropy


def _looks_like_secret_by_entropy(candidate: str) -> bool:
    # Require a mix of letters and digits (pure hex digest-looking strings
    # are common and mostly benign, e.g. git commit hashes) OR high enough
    # entropy regardless of composition. This keeps false positives on
    # things like long words or hashes lower while still being fail-safe.
    has_digit = any(c.isdigit() for c in candidate)
    has_alpha = any(c.isalpha() for c in candidate)
    if not (has_digit and has_alpha):
        return False
    return _shannon_entropy(candidate) >= _ENTROPY_THRESHOLD_BITS_PER_CHAR


def _redact_entropy(text: str) -> str:
    def _sub(match: re.Match) -> str:
        candidate = match.group(0)
        if _looks_like_secret_by_entropy(candidate):
            return PLACEHOLDER.format(category="entropy")
        return candidate

    return _CANDIDATE_TOKEN.sub(_sub, text)


def redact(text: str) -> str:
    """Redact known secret shapes from text. Never raises; returns the
    input unchanged (coerced to str) if it is falsy."""
    if not text:
        return text if text is not None else ""

    result = str(text)

    result = _PEM_PRIVATE_KEY.sub(PLACEHOLDER.format(category="private_key"), result)
    result = _CONN_STRING.sub(PLACEHOLDER.format(category="conn_string"), result)
    result = _HEALTHCHECK_URL.sub(PLACEHOLDER.format(category="healthcheck"), result)
    result = _KNOWN_TOKEN_PREFIXES.sub(PLACEHOLDER.format(category="token"), result)

    def _sub_generic(match: re.Match) -> str:
        return f"{match.group(1)}={PLACEHOLDER.format(category='token')}"

    result = _GENERIC_TOKEN_ASSIGNMENT.sub(_sub_generic, result)
    result = _BEARER_HEADER.sub(
        lambda m: "bearer " + PLACEHOLDER.format(category="token"), result
    )

    result = _PRIVATE_IPV4.sub(PLACEHOLDER.format(category="internal_host"), result)
    result = _INTERNAL_HOSTNAME.sub(PLACEHOLDER.format(category="internal_host"), result)

    result = _redact_entropy(result)

    return result

File: src/OpsConsole.Mcp/Security/Redactor.cs
using System.Text.RegularExpressions;

namespace OpsConsole.Mcp.Security;

/// <summary>
/// Single centralized secret redactor (03-sicurezza-threat-model.md §5/§6). Applied to every
/// outgoing string: tool output, error messages, and audit log arguments/results. Matches are
/// replaced with a fixed placeholder "[REDACTED:&lt;category&gt;]" rather than removed, so the
/// caller can see redaction happened without ever seeing the original value.
/// </summary>
public sealed partial class Redactor
{
    public const string Placeholder = "[REDACTED:";

    // Order matters: more specific patterns run before the generic entropy fallback so we
    // report the most useful category, not just "high-entropy".
    private static readonly (string Category, Regex Pattern)[] Patterns =
    [
        ("pem_private_key", PemBlockRegex()),
        ("known_token", KnownTokenRegex()),
        ("connection_string", ConnectionStringRegex()),
        ("healthcheck_url", HealthcheckUrlRegex()),
        ("generic_secret_kv", GenericSecretKeyValueRegex()),
        ("private_ip", PrivateIpRegex()),
        ("internal_hostname", InternalHostnameRegex()),
    ];

    /// <summary>Redacts all known categories plus a fail-safe entropy heuristic.</summary>
    /// <returns>The redacted string and whether at least one redaction was applied.</returns>
    public (string Text, bool Redacted) Redact(string? input)
    {
        if (string.IsNullOrEmpty(input))
        {
            return (input ?? "", false);
        }

        var text = input;
        var redacted = false;

        foreach (var (category, pattern) in Patterns)
        {
            text = pattern.Replace(text, m =>
            {
                redacted = true;
                return $"{Placeholder}{category}]";
            });
        }

        // Fail-safe: any remaining long, high-entropy, space-free token-like run is redacted
        // even if it matched no known pattern. Prefer a false positive here to a leaked secret.
        text = HighEntropyTokenRegex().Replace(text, m =>
        {
            if (m.Value.Contains(Placeholder))
            {
                return m.Value;
            }

            if (LooksHighEntropy(m.Value))
            {
                redacted = true;
                return $"{Placeholder}entropy]";
            }

            return m.Value;
        });

        return (text, redacted);
    }

    /// <summary>Convenience overload for a batch of lines (e.g. container/journal logs).</summary>
    public IReadOnlyList<string> RedactLines(IEnumerable<string> lines, out int redactedCount)
    {
        var result = new List<string>();
        var count = 0;
        foreach (var line in lines)
        {
            var (text, redacted) = Redact(line);
            result.Add(text);
            if (redacted)
            {
                count++;
            }
        }

        redactedCount = count;
        return result;
    }

    private static bool LooksHighEntropy(string candidate)
    {
        if (candidate.Length < 24)
        {
            return false;
        }

        var hasDigit = false;
        var hasUpper = false;
        var hasLower = false;
        foreach (var c in candidate)
        {
            if (char.IsDigit(c)) hasDigit = true;
            else if (char.IsUpper(c)) hasUpper = true;
            else if (char.IsLower(c)) hasLower = true;
        }

        // Require a mix of classes so ordinary words/sentences are left alone.
        var classCount = (hasDigit ? 1 : 0) + (hasUpper ? 1 : 0) + (hasLower ? 1 : 0);
        return classCount >= 2;
    }

    [GeneratedRegex(@"-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----")]
    private static partial Regex PemBlockRegex();

    // Known provider token prefixes (GitHub PAT/fine-grained, GitHub Actions, Slack, generic AWS).
    [GeneratedRegex(@"\b(ghp_|gho_|ghu_|ghs_|ghr_|github_pat_)[A-Za-z0-9_]{20,}\b|\bAKIA[0-9A-Z]{16}\b|\bxox[baprs]-[A-Za-z0-9-]{10,}\b")]
    private static partial Regex KnownTokenRegex();

    // scheme://user:password@host[:port][/path] — postgres, redis, mysql, amqp, etc.
    [GeneratedRegex(@"\b[a-zA-Z][a-zA-Z0-9+.-]*://[^\s/@:]+:[^\s/@]+@[^\s/]+")]
    private static partial Regex ConnectionStringRegex();

    // Healthcheck ping URLs (e.g. hc-ping.com/<uuid>-style tokens): the path segment itself acts
    // as a bearer credential, so the whole URL is treated as a secret, not just logged as info.
    [GeneratedRegex(@"https?://[^\s]*(?:ping|healthchecks?)[^\s]*/[0-9a-fA-F-]{8,}")]
    private static partial Regex HealthcheckUrlRegex();

    // key=value / header-style secrets: token=, key=, secret=, password=, authorization:, bearer .
    [GeneratedRegex(@"(?i)\b(token|api[_-]?key|secret|password|passwd)\s*[=:]\s*[""']?[A-Za-z0-9_\-.+/=]{8,}[""']?|(?i)\b(authorization)\s*:\s*[""']?bearer\s+[A-Za-z0-9_\-.+/=]{8,}[""']?|(?i)\bbearer\s+[A-Za-z0-9_\-.+/=]{8,}")]
    private static partial Regex GenericSecretKeyValueRegex();

    // RFC1918 private ranges + CGNAT 100.64.0.0/10 (used by mesh VPNs such as Tailscale) + loopback.
    [GeneratedRegex(@"\b(10\.\d{1,3}\.\d{1,3}\.\d{1,3}|172\.(1[6-9]|2\d|3[01])\.\d{1,3}\.\d{1,3}|192\.168\.\d{1,3}\.\d{1,3}|100\.(6[4-9]|[7-9]\d|1[01]\d|12[0-7])\.\d{1,3}\.\d{1,3}|127\.0\.0\.1)\b")]
    private static partial Regex PrivateIpRegex();

    // Internal-looking hostnames: .local / .internal / .lan TLD-style suffixes, or bare
    // "host.tailnet-ish" single-label + reserved suffix combos.
    [GeneratedRegex(@"\b[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.(local|internal|lan|home\.arpa)\b")]
    private static partial Regex InternalHostnameRegex();

    // Fallback token shape: 24+ contiguous non-space alnum/symbol characters.
    [GeneratedRegex(@"[A-Za-z0-9_\-./+=]{24,}")]
    private static partial Regex HighEntropyTokenRegex();
}
